### ISSUE TITLE
RFC: Internal websocket handling

### ISSUE THREAD

**Issue by ahopkins** [sanic-org/sanic#2000]
Sanic uses the `websockets` package to handle the websocket protocol. It has [been discussed](https://github.com/sanic-org/sanic/issues/1908#issuecomment-694172043) that we could remove this dependency and implement them ourselves. More info to come.

**Comment by aaugustin** [comment]
`websockets` now has a Sans-I/O interface, which I designed pretty much for your use case. It removes all buffering — give it some bytes, get the next message back. You may want to evaluate it before building your own from scratch!

You will still have to implement a keepalive with ping/pong, message fragmentation / defragmentation, etc. if you want these features, because they cannot be provided in a Sans-I/O layer.

Right now I'm working on integrating the Sans-I/O implementation in a high-level interface based on sockets + threads (which makes sense for clients, not for servers). I can push that code in a WIP MR if you want to get a feeling for what the result looks like.

Of course you can also build a WebSocket implementation from scratch, it's fun and not all that hard!

**Comment by ahopkins** [comment]
@aaugustin Thanks for heads up. It is not an issue of _fun_ (although, I don't doubt that you are correct). 

The issue has for us has largely been a matter of support. We have tried to remove third party libraries because we get stuck in a pinned version when underlying APIs changes. With regards to your project, for example, this happened with v7 that we were stuck with for a long time. Ultimately this is a bad experience for the developer. 

The other issue is that we just overhauled the io to effectively make all of Sanic one big streaming layer.

Your new implementation sounds very interesting. If there is a branch to follow the code, then I am not so worried about the documentation. Have you thought about splitting the package into layers? This is what the encode project did with httpx (and httpcore).

**Comment by aaugustin** [comment]
Yes, the package is now split into layers.

`websockets`contains the new, Sans-I/O layer. To use it:

- Initialize a [ServerConnection](https://github.com/aaugustin/websockets/blob/aa93c4ceca90a1798f86b2fc2b110a42f308d721/src/websockets/server.py#L59)
- When you get data from the network, push it with [receive_data](https://github.com/aaugustin/websockets/blob/aa93c4ceca90a1798f86b2fc2b110a42f308d721/src/websockets/connection.py#L123) or [receive_eof](https://github.com/aaugustin/websockets/blob/aa93c4ceca90a1798f86b2fc2b110a42f308d721/src/websockets/connection.py#L136), then call [events_received](https://github.com/aaugustin/websockets/blob/aa93c4ceca90a1798f86b2fc2b110a42f308d721/src/websockets/connection.py#L218) to recover any events produced by the connection (handshake request or websocket frame) and [data_to_send](https://github.com/aaugustin/websockets/blob/aa93c4ceca90a1798f86b2fc2b110a42f308d721/src/websockets/connection.py#L231) to get data to send to the client
- Call any of the [send_*](https://github.com/aaugustin/websockets/blob/aa93c4ceca90a1798f86b2fc2b110a42f308d721/src/websockets/connection.py#L162) methods to send a frame, then call [data_to_send](https://github.com/aaugustin/websockets/blob/aa93c4ceca90a1798f86b2fc2b110a42f308d721/src/websockets/connection.py#L231) to get data to send to the client

Pros: I'm strongly confident that the Sans-I/O implementation is robust because I've been working on the WebSocket protocol for 7 years now, I [spent time designing the API](https://fractalideas.com/blog/sans-io-when-rubber-meets-road/), and I a solid test suite.

Cons: the API is somewhat cumbersome vs. the corountine-based approach of the previous implementation, but it provides the benefits of Sans-I/O. Also the error handling, especially on premature EOF, could still change depending on how the integration with I/O works.

Then you still have to write something of this order of magnitude: https://github.com/aaugustin/websockets/pull/885/files (only the golden path works, no error handling yet).

----

`websockets.legacy` contains the historical, asyncio-based implementation, which Sanic integrated in the best way you could, given the lack of a public API. I saw the problems it caused and that was a significant driver for building the Sans-I/O layer.

My plan is to build `websockets.sync`, `websockets.asyncio`, and maybe others.

**Comment by aaugustin** [comment]
For the avoidance of doubt: the Sans-I/O implementation is:

- merged to the master branch of websockets;
- not released yet, but I was planning to make a release pretty much in the current state of the repository.

**Comment by ahopkins** [comment]
Thank you @aaugustin. It seems this is very good timing.

**Comment by ashleysommer** [comment]
Some updates since January:

`websockets` module v9.0 and 9.1 are released with the Sans-IO code. This is a breaking update, because existing websocket protocol functionality has moved to `websockets.legacy` submodule. Some projects are moving over to the new version, but a lot are pinning to `websockets<9` in order to avoid having to deal with making the transition (to either Sans-IO or legacy).

Sanic already has its own robust (and _fast_) I/O layer, so it makes sense for Sanic to use the new Sans-IO features of `websockets` and attach that to our existing I/O, and provide a custom interface over that for Sanic applications to utilize.

I've made some progress in a local branch.. after a bumpy start, things are now working pretty well.

The `WebsocketProtocol` class in `websockets.py` is already a subclass of `HttpProtocol`. That means it has access to the underlying streaming transport layer, along with niceties like backpressure-controlled response-streaming etc.

The big change is the removal of the `websockets.WebsocketCommonProtocol` class (it is now a legacy feature). Some of the features of that are now provided by the `websockets.connection.ServerConnection` and `Connection` classes but some needs to be implemented by Sanic (as expected).
The Sanic `WebsocketProtocol` class itself grows to fill that gap.

Quite a lot of the new code in Sanic `WebsocketProtocol` was stolen (borrowed?) directly from the reference implementation provided [here](https://github.com/aaugustin/websockets/pull/885/files). That reference implementation is completely synchronous (and thread-based), so I've converted it all to asyncio-based patterns, it seems to be working well.
For backward compatibility and common-api-surface reasons, I've also ported a number of features from `legacy.WebsocketCommonProtocol` into `WebsocketProtocol`.

Some hiccups:

The biggest issue I had initially was caused by `websockets.ServerConnection` to do the websocket handshake via the `accept()` method. What I didn't realize at the time is `ServerConnection` implements its own minimal HTTP request/response cycle, which reads in a HTTP request, checks if it is an upgrade request, executes the upgrade handshake, and sends an upgrade response, before switching over to websocket operating mode.

That didn't work in Sanic's case because all traffic flows through the `HttpProtocol` and `app.handle_request`, and the router first, before arriving at the websocket handler. If you try to execute `accept` at this point then `websockets.ServerConnection` cannot re-read a Request from the transport. Alternatively if we manually execute our own handshake and upgrade the connection, then when we attempt to use `websockets.ServerConnection` to receive the process websocket data it crashes because the `serverConnection` instance is still waiting to receive a HTTP request (it hasn't transitioned to websocket operating mode).

The solution was to initially use a `websockets.ServerConnection` instance to help generate parts of our own handshake routine, then create a `websockets.Connection` object, copy the properties from `ServerConnection` to `Connection`, and use that as the websocket handler going forward. That works surprisingly smoothly.

I've still got some more work to go on it. There are a lot of synchronization primitives in the Sync Server reference implementation which need to be modified or simply removed in our asyncio-version.

**Comment by aaugustin** [comment]
Hello!

Thanks for reporting your progress! You're moving faster than me so supporting you is now my top priority for the Sans-I/O layer in websockets!

It's nice of you to call [my prototype](https://github.com/aaugustin/websockets/pull/885) a "reference implementation". To be honest, it's rather "me trying to figure out how these things work":-) I started with sockets & threads because I already have a good asyncio implementation (the one now called legacy) and I wanted to meet the demand for writing synchronous clients.

----

Here's how I expected the Sans-I/O layer to be used by projects like Sanic. The main issue I saw — and indeed one you bumped into — is:

- both sanic and websockets provide a HTTP implementation
- you want to use the HTTP implementation from sanic and the WebSocket handshake negotiation from websockets

I had two solutions in mind. You found a third one; it looks good to me.

**Solution 1**

Re-serialize the HTTP request and push it to websockets, who will parse it again.

This doesn't feel like great software engineering, but it would work. It's pretty much what a proxy does.

Given sanic's focus on performance, I don't expect you to like this :-)

**Solution 2**

Override `ServerConnection.parse` so it does the same as `Connection.parse` — ignoring the handshake and parsing only WebSocket frames.

Then:

- you parse the HTTP request as usual in sanic
- if you determine it's a WebSocket Upgrade request, you initialize a `ServerConnection` (with `parse` overridden), you call its `accept` method with an object mimicking a `websockets.Request` and you send the response if the handshake is successful
- then you take any data that may still be buffered in sanic and push it to the websockets connection, and you push any new data to the websockets connection

This relies on some private implementation details, but seems manageable.

----

The Sans-I/O layer doesn't do things like "reassemble frames of fragmented messages" or "send pings every N seconds and timeout if you don't get a pong". Such features depend on a specific concurrency model (e.g. asyncio) so the Sans-I/O layer isn't allowed to do them.

To see the whole kit of parts you could reuse, you should look not only at my prototype for integrating the Sans-I/O layer, but also at the current `WebSocketCommonProtocol` class (keeping in mind that it carries quite a bit of legacy... which is why I want to rewrite it...)

----

Is the Sans-I/O layer stable? I don't expect major changes. I'm still unsure how to communicate that the connection was closed on must be closed between the Sans-I/O and the I/O layers. Happy to get feedback in this area, or anywhere else!

----

Aside -- I feel bad about "a lot are pinning to websockets<9" and it hurts to read "breaking update" after putting a lot of effort into backwards compatibility (vs. doing a websockets2 for example).

Some projects ran into trouble because they restrict themselves the subset of Python that mypy understands, so their static typing checks failed, but that was resolved within a few days after the release of 9.0.

There's a hard situation with uvicorn who depends on a private API. I agreed to continue supporting them for at least five years. I feel like this is a good resolution  for them — usually, if you use private APIs, you're on your own...

If there are other issues, I'll do my best to find a good resolution!

**Comment by ahopkins** [comment]
Thanks @ashleysommer @aaugustin. Great work, and lots of good information here. I'm pretty sure that's exactly where I gave up last time I tried working on this 😆.

**Comment by aaugustin** [comment]
Hello @ashleysommer, would https://github.com/aaugustin/websockets/pull/978 help websockets attempting to re-read the handshake request from the transport?

It should make it possible to handle the handshake like this:

```python

from websockets.connection import State
from websockets.http11 import Request
from websockets.server import ServerConnection

# Here Sanic parses the HTTP request...

connection = ServerConnection(state=State.OPEN)
# Convert sanic's request into a websockets request
# I didn't check how Sanic handles headers, maybe you need more conversion
ws_request = Request(request.raw_url, request.headers)
response = connection.accept(ws_request)

# Here you send the response as you like, like with Sanic's stack

if response.status_code != 101:
    ... # abort and close connection
```

**Comment by aaugustin** [comment]
Quick note on API stability: when the request is invalid, right now, websockets smuggles the resulting exception by attaching it to the response object. I dislike this design and may change it. This is why I'm testing for `if response.status_code != 101` rather than `if request.exception is None`.

**Comment by ashleysommer** [comment]
Hi @aaugustin 
> would aaugustin/websockets#978 help with websockets attempting to re-read the handshake request from the transport?
Yes, this looks great! That would allow me to remove the little `ServerConnection`->`Connection` switcheroo I have in the code.

**Comment by aaugustin** [comment]
Cool. I'll merge it.

I expect we may need a few other changes like this one. If your dev setup can accommodate it, it would be best to work against websockets' master branch. Then, when you need to make a release, I'll make one before — either by releasing master or by backporting everything you need to a 9.x branch.

**Comment by ashleysommer** [comment]
Yep, sounds good. I switch my dev env over to `master`.

**Comment by Tronic** [comment]
Btw, would it be possible to use the sans-i/o with external buffers? We have a `recv_buffer` which is a `bytearray`. Protocols can use the buffer and `del recv_buffer[:bytes_used]` to consume data, or call `await receive_more()` to wait for another incoming packet, which gets appended to the buffer. Protocols are expected to leave any partial messages or otherwise unused data in the buffer.

This is done to avoid copying, for maximal performance. Naturally Sanic is concerned with being very fast, so it would be good if this copying was avoided with websockets as well.

**Comment by ashleysommer** [comment]
Thats a good question. The sans-io `websockets.Connection` implementation doesn't buffer much really. It has a tiny buffer that it uses while building a Websocket Frame. As soon as a frame is complete, it emits it as an Event. If the event's frame is part of a Websocket Message, that frame is passed to a Websocket Message Assembler. The assembler buffers the frames until a full message is built, then emits the message.

In my current working implementation, `data_received()` handler on `WebsocketProtocol` passes the incoming data directly to the sans-io `websockets.Conneciton`, it doesn't go into the HttpProtocol recv_buffer. So there is no duplication of received data.

Having said that, yes it _may_ be possible (with the help of @aaugustin) to switch to using our existing Sanic receive buffer at the HttpProtocol level.

**Comment by aaugustin** [comment]
Indeed, going straight to [websockets's buffer](https://github.com/aaugustin/websockets/blob/6722857ebefd365c9299a79e1526644e25fe1015/src/websockets/streams.py#L15) instead of sanic's buffer should have the same performance characteristics, given that both are implemented using `bytearray`.

FYI:

- I tried another buffer implementation based on `collections.deque(bytes)` to optimize performance but it didn't yield measurable gains.
- I've been thinking about avoiding copies because I wanted to get as close to zero-copy as I could get. Some copies are hard to avoid:
    - decoding bytes to str makes a copy
    - masking/unmasking makes a copy — it could be done in place but I'm not convinced it would be faster

Surely I could make optimizations; however, it's not like I've been ignoring performance entirely :-)

(In the same area, I pioneered the [XOR with SSE2 instructions](https://github.com/aaugustin/websockets/blob/6722857ebefd365c9299a79e1526644e25fe1015/src/websockets/speedups.c#L133-L145) approach, which most other Python implementations of WebSockets subsequently copied.)

**Comment by Tronic** [comment]
@aaugustin Receiving from socket also makes a `bytes` object that is hard to avoid (I haven't seen anyone using asyncio's `sock_recv_into` and trio doesn't implement it at all). Just checked some of your code -- great work with your optimizations! I also tried list(bytes) and other variations, and looks like we ended up with the same result :)

We'd also prefer being able to do streaming so that the webapp's call `await ws.receive()` internally uses `await receive_more()` until a message can be returned, possibly as a view to that same bytearray (no copying, again). This *pull architecture* is how we avoid buffer queues in normal HTTP streaming, and avoid ever having to return with "need more data".

@ashleysommer Good that at least that extra buffering is then avoided. Some profiling or benchmarking would be useful to find out how much all of this actually matters. Just making my remarks here because working Sanic into that pull architecture was a huge undertaking but it had also very clear benefits in both performance and simplicity.

**Comment by aaugustin** [comment]
Yes, there's an annoying incompatibility between:

1. Streaming via coroutines = this sort of data flow: https://websockets.readthedocs.io/en/stable/design.html#data-flow
2. Abiding by the rules of Sans-I/O = building something that doesn't use any async / concurrency primitives so it's easy to integrate anywhere else.

I started websockets with the first approach and I'm now redoing it with the second approach. I agree that the first approach provides great performance and simplicity within a single piece of software. However, it's hard to integrate (see #1214). The main benefits of the second one are easy integration in various contexts and easy testing.

----

Not buffering a single byte unnecessarily is a design goal of the Sans-I/O refactor. The end result in terms of how much data is in flight is identical to the streaming scenario, like @ashleysommer explained, if you take care of fetching events right after pushing data.

With what you have in mind, this happens (-> = push, <- = pull)

network -> OS -> event loop -> `recv_buffer` <- stream reader <- coroutines all the way to the public API

With what Ashley is building, I believe this happens:

network -> OS -> event loop -> websockets -> message <- coroutines all the way to the public API

In other words, the push -> pull switch (which requires backpressure to work correctly) happens at the level of the stream of messages rather than at the level of the bytestream. This requires thinking a bit about backpressure.

Note that you aren't seeing the dreaded queue of incoming messages here. It's legacy dating back to before I realized I needed a separate coroutine just for managing the connection. It's going away. At least you're gaining that from the move.

(BTW, you kept WEBSOCKET_MAX_QUEUE = 32. Until you remove it, I think you could set it to 1. It's generally useless, except when you have a slow consumer that cannot keep up with the stream of messages and you don't want to time out.)

----

websockets does use a view of the underlying bytearray like you suggest :-)

Then there's further processing:

- Messages are masked in the flow from client to server. You need to unmask them.
- You need to decompress them (unless compression is disabled).
- You need to `bytes.decode()` them (unless it's a binary frame).

So, at the end, you don't get a bytearray, but one copy was saved at the first step of this process!

**Comment by aaugustin** [comment]
@ashleysommer Just wanted to let you know about https://github.com/aaugustin/websockets/pull/979 which changes `Frame` to a `dataclass` and also changes the order of fields. You shouldn't be affected, but on the off chance you are, here's how to adapt.

The `Frame` events returned by `Connection.events_received` changed. If you only access their attributes, you're OK. If you relied on the fact that they were tuples e.g. `fin, opcode, data, rsv1, rsv2, rsv3 = frame` then you need to move `fin` between `data` and `rsv1'.

The `Connection.send_*` methods didn't change -- they already had the target API.

**Comment by ashleysommer** [comment]
@aaugustin 
Thanks for your insights and your ongoing help with this.

>With what Ashley is building, I believe this happens:
>network -> OS -> event loop -> websockets -> message <- coroutines all the way to the public API

Yes, that's right. As you mentioned, when integrating the push-based asyncio->websockets model with the pull-based Sanic application model, there will always need to be a meeting point. And in my working implementation that point is the Message Assembler (I've borrowed/stolen the pattern from your example Assembler from here: https://github.com/aaugustin/websockets/blob/6eb98dd8fa5b2c896b9f6be7e8d117708da82a39/src/websockets/sync/messages.py).

>BTW, you kept WEBSOCKET_MAX_QUEUE = 32. Until you remove it, I think you could set it to 1.

I have no message queue implemented right now. I was debating whether there should be one at all. In the current implementation the Message Assembler can receive (concat) frames until a full Message is assembled, then it sets the `message_complete` event, until that message is Pulled (consumed) by the application. This means any incoming new frames also get blocked from being added to the assembler, until the message is cleared. One thing I could add to this logic is initiating a `transport.pause_reading()` right as `message_complete` is set, and `transport.resume_reading()` when the message is cleared. That would mean the incoming websocket data are blocked at the TCP level, rather than the message-assembler level (potentially faster, cleaner, "more correct").

Another idea I was toying with the ability to maybe have a completed-message queue of 1, so that a single message can be waiting to be pulled by the application while the next one is being received+assembled. That would remove a delay at the application level when a single message is finished being processed then application asks for the next one (the next one is already assembled and ready to go).

> Just wanted to let you know about aaugustin/websockets#979 which changes Frame to a dataclass and also changes the order of fields. You shouldn't be affected, but on the off chance you are, here's how to adapt.

That's fine, its a good change. I don't think there's anything I need to change in my code to accommodate that.

> going straight to websockets's buffer instead of sanic's buffer

I noticed the size of the `websockets` StreamReader buffer is unlimited. Is that intentional? I was reading [this article](http://www.lenholgate.com/blog/2011/07/websockets-is-a-stream-not-a-message-based-protocol.html), I know its pretty old now, but it mentions that in theory a websocket frame can be 9,223,372,036,854,775,807 bytes (2^63) in size. Obviously python would crash long before receiving a frame of that magnitude, but could there be an internal size limit (maybe 1MB?) on the StreamReader buffer, when it reaches that it blocks pushing more data in until some content is read out first?

**Comment by aaugustin** [comment]
> I have no message queue implemented right now.

My statement about the message queue was referring to the current implementation of sanic, not to your development branch. Apologies for the confusion!

> Another idea I was toying with the ability to maybe have a completed-message queue of 1

A queue of 1 would create the possibility for a little bit of unnecessary buffering so I wouldn't do it, unless it simplifies the implementation drastically.

> I noticed the size of the websockets StreamReader buffer is unlimited. Is that intentional? 

There's a `max_size` on each frame, so as long as you read any complete frame after pushing data and before pushing even more data, you're good. (I think.)

**Comment by ashleysommer** [comment]
@aaugustin @ahopkins 
I've pushed my new working implementionation into a dev branch: https://github.com/ashleysommer/sanic/tree/new_websockets

I've spent another couple of days on it, and its now passing all of the tests, including some other demo applications I've written to stress it out.

It certainly needs cleaning up, and could probably get some efficiency improvements in places too.

That branch also incorporates another change I've written, that splits the existing `server.HttpProtocol` out into a base `SanicProtocol` class, that contains just all of the connection-handling and transport-layer stuff, and `HttpProtocol` that is a subclass of `SanicProtocol` and adds the HTTP-specific handlers on top of that. I'd like the new `WebSocket` protocol to eventually subclass `SanicProtocol` but for now it still subclasses `HttpProtocol`.
